ml4gw/utils/slicing.py

```diff
@@ -13,6 +14,67 @@
 ]
 
 
+def unfold_windows(
+    x: torch.Tensor,
+    window_size: int,
+    stride: int,
+    drop_last: bool = True,
+):
+    """Unfold a timeseries into windows
+
+    Args:
+        x:
+            The timeseries to unfold. Can have shape
+            `(batch_size, num_channels, length * sample_rate)`,
+            `(num_channels, length * sample_rate)`, or
+            `(length * sample_rate)`
+        window_size:
+            The size of the windows to unfold from x
+        stride:
+            The stride between windows
+        drop_last:
+            If true, does not return the remainder that exists
+            when the timeseries cannot be evenly broken up into
+            windows
+
+    Returns:
+       A tensor of shape
+       `(num_windows, batch_size, num_channels, kernel_size)`,
+       `(num_windows, num_channels, kernel_size)`, or
+       `(num_windows, kernel_size)` depending on whether the
+       input tensor is 3D, 2D, or 1D
+
+       If `drop_last` is false, returns the remainder of the
+       timeseries, shaped to be compatible with the returned
+       unfolded tensor
+    """
+    reshape = list(x.shape[:-1])
+    if x.ndim == 1:
+        x = x[None, None, None, :]
+    elif x.ndim == 2:
+        x = x[None, :, None, :]
+    elif x.ndim == 3:
+        x = x[:, :, None, :]
+
+    num_windows = (x.shape[-1] - window_size) // stride + 1
```

Comment:
I think you can do this calc before the reshape, since the last axis doesn't change, and can slice out the remainder here while it already has the shape that you want. Something like:

```python
num_windows = (x.shape[-1] - window_size) // stride + 1
remainder = x.shape[-1] - window_size - num_windows * stride
if remainder == 0:
    if not drop_last:
        # create an empty tensor for consistency
        shape = list(x.shape)
        shape[-1] = 0
        remainder = torch.zeros(shape, dtype=x.dtype, device=x.device)
else:
    # separate x from its remainder _regardless_ of whether
    # we end up returning it or not
    x, remainder = torch.split(x, [x.shape[-1] - remainder, remainder], dim=-1)
```
Then at the end you can just do
```python
if not drop_last:
    return x, remainder
return x
```

Reply:
So we want `remainder` to have the same dimensionality as the input tensor, not the output? Or should I still reshape it before returning?

Reply:
Hmmm that's a good question. I was thinking yes yesterday, but maybe it does make sense to prepend a dummy dimension for the "one window" this constitutes. I guess it depends on what the most common use cases for the last bit are, and that will come with time. Probably not worth agonizing over so I'll leave it up to you, whatever you think is more consistent I'm cool with